extensions/amp-analytics/0.1/analytics-root.js

```diff
@@ -0,0 +1,317 @@
+      return found;
+    }
+    return null;
+  }
+
+  /**
+   * Creates listener-filter for DOM events to check against the specified
+   * selector. If the node (or its ancestors) match the selector the listener
+   * will be called.
+   *
+   * @param {function(!Element, !Event)} listener The first argument is the
+   *   matched target node and the second is the original event.
+   * @param {!Element} context
+   * @param {string} selector DOM query selector.
+   * @param {?string=} selectionMethod Allowed values are `null`,
+   *   `'closest'` and `'scope'`.
+   * @return {function(!Event)}
+   */
+  createSelectiveListener(
+      listener, context, selector, selectionMethod = null) {
+    return event => {
+      if (selector == ':host') {
+        // `:host` is not reachable via selective listener b/c event path
+        // cannot be retargeted across the boundary of the embed.
+        return;
+      }
+
+      // Navigate up the DOM tree to find the actual target.
+      const rootElement = this.getRootElement();
+      const isSelectAny = (selector == '*');
+      const isSelectRoot = (selector == ':root');
+      let target = event.target;
+      while (target) {
+
+        // Target must be contained by this root.
+        if (!this.contains(target)) {
+          break;
+        }
+        // `:scope` context must contain the target.
+        if (selectionMethod == 'scope' &&
+            !isSelectRoot && !context.contains(target)) {
+          break;
+        }
+        // `closest()` target must contain the conext.
+        if (selectionMethod == 'closest' && !target.contains(context)) {
+          // However, the search must continue!
+          target = target.parentElement;
+          continue;
+        }
+
+        // Check if the target matches the selector.
+        if (isSelectAny ||
+            isSelectRoot && target == rootElement ||
+            matchesNoInline(target, selector)) {
+          listener(target, event);
+          // Don't fire the event multiple times even if the more than one
+          // ancestor matches the selector.
+          break;
+        }
+
+        target = target.parentElement;
+      }
+    };
+  }
+}
+
+
+/**
+ * The implementation of the analytics root for an ampdoc.
+ */
+export class AmpdocAnalyticsRoot extends AnalyticsRoot {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    super(ampdoc, /* parent */ null);
+  }
+
+  /** @override */
+  getType() {
+    return 'ampdoc';
+  }
+
+  /** @override */
+  getRoot() {
+    return this.ampdoc.getRootNode();
+  }
+
+  /** @override */
+  getHostElement() {
+    // ampdoc is always the root of everything - no host.
+    return null;
+  }
+
+  /** @override */
+  getElementById(id) {
+    return this.ampdoc.getElementById(id);
+  }
+}
+
+
+/**
+ * The implementation of the analytics root for FIE.
+ */
+export class EmbedAnalyticsRoot extends AnalyticsRoot {
```

Comment:
No. ShadowDOM is covered by the root. If we actually go back to shadow embeds - we can worry about this then.